torch/nn/modules/transformer.py

```diff
@@ -0,0 +1,519 @@
+        self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.ff = FeedForward(d_model, d_ff=d_ff, dropout=dropout)
+        self.norm1 = LayerNorm(d_model)
+        self.norm2 = LayerNorm(d_model)
+        self.dropout1 = Dropout(dropout)
+        self.dropout2 = Dropout(dropout)
+
+    def forward(self, src, src_mask=None):
+        r"""Pass the input through the endocder layer.
+
+        Args:
+            src: the sequnce to the encoder layer (required).
+            src_mask: the mask for the src sequence (optional).
+
+        Shape:
+            see the docs in Transformer class.
+        """
+        src2 = self.norm1(src)
+        src = src + self.dropout1(self.self_attn(src2, src2, src2, attn_mask=src_mask)[0])
+        src2 = self.norm2(src)
+        src = src + self.dropout2(self.ff(src2))
+        return src
+
+
+class TransformerDecoderLayer(Module):
+    r"""TransformerDecoderLayer is made up of self-attn, multi-head-attn and feedforward network.
+
+    Args:
+        d_model: the embed dim (required).
+        nhead: the number of heads in the multiheadattention models (required).
+        d_ff: the dimension of the feedforward network model (default=2048).
+        dropout: the dropout value (default=0.1).
+
+    Examples::
+        >>> decoder_layer = nn.TransformerDecoderLayer(d_model, nhead)
+    """
+
+    def __init__(self, d_model, nhead, d_ff=2048, dropout=0.1):
+        super(TransformerDecoderLayer, self).__init__()
+        self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.multihead_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.ff = FeedForward(d_model, d_ff=d_ff, dropout=dropout)
+        self.norm1 = LayerNorm(d_model)
+        self.norm2 = LayerNorm(d_model)
+        self.norm3 = LayerNorm(d_model)
+        self.dropout1 = Dropout(dropout)
+        self.dropout2 = Dropout(dropout)
+        self.dropout3 = Dropout(dropout)
+
+    def forward(self, tgt, memory, tgt_mask=None, memory_mask=None):
+        r"""Pass the inputs (and mask) through the decoder layer.
+
+        Args:
+            tgt: the sequence to the decoder layer (required).
+            memory: the sequnce from the last layer of the encoder (required).
+            tgt_mask: the mask for the tgt sequence (optional).
+            memory_mask: the mask for the memory sequence (optional).
+
+        Shape:
+            see the docs in Transformer class.
+        """
+        tgt2 = self.norm1(tgt)
+        tgt = tgt + self.dropout1(self.self_attn(tgt2, tgt2, tgt2, attn_mask=tgt_mask)[0])
+        tgt2 = self.norm2(tgt)
+        tgt = tgt + self.dropout2(self.multihead_attn(tgt2, memory, memory, attn_mask=memory_mask)[0])
+        tgt2 = self.norm3(tgt)
+        tgt = tgt + self.dropout3(self.ff(tgt2))        
+        return tgt
+
+
+# Temporarily leave LayerNorm module here. Will be moved somewhere else.
+class LayerNorm(Module):
```

Comment:
This should be replaceable by a [normalization layer](https://pytorch.org/docs/stable/nn.html#batchnorm1d).

Reply:
we have a native LayerNorm as well... https://pytorch.org/docs/stable/nn.html?highlight=layernorm#torch.nn.LayerNorm